File: app/src/main/java/com/tapan/weather/forecast/home/presenter/HomePresenter.java
package com.tapan.weather.forecast.home.presenter;

import android.util.Log;
import com.tapan.weather.forecast.R;
import com.tapan.weather.forecast.core.base.presenter.BasePresenter;
import com.tapan.weather.forecast.core.network.callback.WeatherAPIHandler;
import com.tapan.weather.forecast.core.network.request.WeatherRequest;
import com.tapan.weather.forecast.core.network.response.WeatherResponse;
import com.tapan.weather.forecast.core.rx.BaseObserver;
import com.tapan.weather.forecast.core.util.CoreConstants;
import com.tapan.weather.forecast.core.util.CoreUtils;
import com.tapan.weather.forecast.home.callback.HomeListener;
import com.tapan.weather.forecast.home.persistence.GeoLocation;
import com.tapan.weather.forecast.home.view.HomeView;
import io.reactivex.CompletableObserver;
import io.reactivex.SingleObserver;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.internal.disposables.DisposableHelper;
import io.reactivex.schedulers.Schedulers;

public class HomePresenter<V extends HomeView> extends BasePresenter<V>
    implements HomeListener<V> {

  private static final String TAG = "HomePresenter";

  public HomePresenter() {
  }

  @Override
  public void getWeatherContent(double latitude, double longitude) {
    getView().getLocationDB().getGeoLocationDao()
        .getLocation(latitude, longitude)
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(weatherCacheObserver(latitude, longitude));
  }

  private SingleObserver<? super GeoLocation> weatherCacheObserver(final double latitude,
      final double longitude) {
    return new BaseObserver<GeoLocation>() {
      @Override
      public void onSuccess(GeoLocation response) {
        getView().hideLoading();
        lazySet(DisposableHelper.DISPOSED);
        getView().updateWeatherDetails(response);
      }

      @Override
      public void onError(Throwable ex) {
        //getView().hideLoading();
        Log.e(TAG, ex.getMessage());
        onWeatherAPICall(latitude, longitude);
      }
    };
  }

  private void onWeatherAPICall(double latitude, double longitude) {
    if (getView() != null) {
      if (getView().isNetworkConnected()) {
        processWeatherContent(latitude, longitude);
      } else {
        getView().showError(R.string.msg_network_error);
      }
    }
  }

  /**
   * Retrieve the weather data
   */
  private void processWeatherContent(double latitude, double longitude) {
    getView().showLoading();
    WeatherAPIHandler weatherAPIHandler = new WeatherAPIHandler(CoreConstants.BASE_URL);

    WeatherRequest request = new WeatherRequest.Builder()
        .setLatitude(latitude)
        .setLongitude(longitude)
        .build();
    weatherContentRequest(weatherAPIHandler, request);
  }

  /**
   * Subscribing the weather content
   *
   * @param weatherAPIHandler API handler
   */
  private void weatherContentRequest(WeatherAPIHandler weatherAPIHandler, WeatherRequest request) {
    weatherAPIHandler.getWeatherDetails(request, CoreConstants.APP_ID)
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(weatherContentObserver());
  }

  /**
   * Handle the success/error response of weather content
   *
   * @return Single observer of result
   */
  private SingleObserver<? super WeatherResponse> weatherContentObserver() {
    return new BaseObserver<WeatherResponse>() {
      @Override
      public void onSuccess(WeatherResponse response) {
        //getView().hideLoading();
        lazySet(DisposableHelper.DISPOSED);
        updatePersistence(response);
      }

      @Override
      public void onError(Throwable ex) {
        getView().hideLoading();
        Log.e(TAG, ex.getMessage());
      }
    };
  }

  private void updatePersistence(WeatherResponse response) {
    GeoLocation geoLocation = new GeoLocation();
    geoLocation.setLatitude(getView().getCurrentLatitude());
    geoLocation.setLongitude(getView().getCurrentLongitude());

    if (response.getMain() != null) {
      geoLocation.setTemperature(CoreUtils.getFormattedCurrentTemp(response.getMain().getTemp()));
      geoLocation.setOverAllTemperature(
          CoreUtils.getFormattedOverallTemp(response.getMain().getTempMin(),
              response.getMain().getTempMax()));
    }

    if (response.getWeather() != null && response.getWeather().size() > 0
        && response.getWind() != null) {
      geoLocation.setDescription(
          CoreUtils.getFormattedDescription(response.getWeather().get(0).getDescription(),
              response.getWind().getSpeed()));
    }

    insertLocationData(geoLocation);
  }

  private void insertLocationData(final GeoLocation geoLocation) {
    getView().getLocationDB().getGeoLocationDao()
        .insertLocation(geoLocation)
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(new CompletableObserver() {
          @Override
          public void onSubscribe(Disposable d) {
          }

          @Override
          public void onComplete() {
            getView().hideLoading();
            getView().updateWeatherDetails(geoLocation);
          }

          @Override
          public void onError(Throwable e) {
            getView().hideLoading();
            getView().updateWeatherDetails(geoLocation);
          }
        });
  }
}

File: app/src/main/java/com/tapan/weather/forecast/home/presenter/HomePresenter.kt
package com.tapan.weather.forecast.home.presenter

import android.util.Log
import com.tapan.weather.forecast.R.string
import com.tapan.weather.forecast.core.base.presenter.BasePresenter
import com.tapan.weather.forecast.core.network.callback.WeatherAPIHandler
import com.tapan.weather.forecast.core.network.request.WeatherRequest
import com.tapan.weather.forecast.core.network.request.WeatherRequest.Builder
import com.tapan.weather.forecast.core.network.response.WeatherResponse
import com.tapan.weather.forecast.core.rx.BaseObserver
import com.tapan.weather.forecast.core.util.CoreConstants
import com.tapan.weather.forecast.core.util.CoreUtils.getFormattedCurrentTemp
import com.tapan.weather.forecast.core.util.CoreUtils.getFormattedDescription
import com.tapan.weather.forecast.core.util.CoreUtils.getFormattedOverallTemp
import com.tapan.weather.forecast.home.callback.HomeListener
import com.tapan.weather.forecast.home.persistence.GeoLocation
import com.tapan.weather.forecast.home.view.HomeView
import io.reactivex.CompletableObserver
import io.reactivex.SingleObserver
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.internal.disposables.DisposableHelper.DISPOSED
import io.reactivex.schedulers.Schedulers

class HomePresenter<V : HomeView> : BasePresenter<V>(), HomeListener<V> {

  companion object {
    private const val TAG = "HomePresenter"
  }

  override
  fun getWeatherContent(
    latitude: Double,
    longitude: Double
  ) {
    view!!.locationDB?.geoLocationDao
        ?.getLocation(latitude, longitude)
        ?.subscribeOn(Schedulers.io())
        ?.observeOn(AndroidSchedulers.mainThread())
        ?.subscribe(weatherCacheObserver(latitude, longitude))
  }

  private fun weatherCacheObserver(
    latitude: Double,
    longitude: Double
  ):
      SingleObserver<in GeoLocation> {
    return object : BaseObserver<GeoLocation?>() {

      override
      fun onSuccess(response: GeoLocation) {
        view!!.hideLoading()
        lazySet(DISPOSED)
        view!!.updateWeatherDetails(response)
      }

      override
      fun onError(ex: Throwable) {
        //getView().hideLoading();
        Log.e(TAG, ex.message)
        onWeatherAPICall(latitude, longitude)
      }
    }
  }

  private fun onWeatherAPICall(
    latitude: Double,
    longitude: Double
  ) {
    if (view != null) {
      if (view!!.isNetworkConnected()) {
        processWeatherContent(latitude, longitude)
      } else {
        view!!.showError(string.msg_network_error)
      }
    }
  }

  /**
   * Retrieve the weather data
   */
  private fun processWeatherContent(
    latitude: Double,
    longitude: Double
  ) {
    view!!.showLoading()
    val weatherAPIHandler = WeatherAPIHandler(CoreConstants.BASE_URL)
    val request = Builder()
        .setLatitude(latitude)
        .setLongitude(longitude)
        .build()
    weatherContentRequest(weatherAPIHandler, request)
  }

  /**
   * Subscribing the weather content
   *
   * @param weatherAPIHandler API handler
   */
  private fun weatherContentRequest(
    weatherAPIHandler: WeatherAPIHandler,
    request: WeatherRequest
  ) {
    weatherAPIHandler.getWeatherDetails(request, CoreConstants.APP_ID)
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(weatherContentObserver())
  }

  /**
   * Handle the success/error response of weather content
   *
   * @return Single observer of result
   */
  private fun weatherContentObserver(): SingleObserver<in WeatherResponse> {
    return object : BaseObserver<WeatherResponse?>() {

      override
      fun onSuccess(response: WeatherResponse) {
        //getView().hideLoading();
        lazySet(DISPOSED)
        updatePersistence(response)
      }

      override
      fun onError(ex: Throwable) {
        view!!.hideLoading()
        Log.e(TAG, ex.message)
      }
    }
  }

  private fun updatePersistence(response: WeatherResponse) {
    val geoLocation = GeoLocation()
    geoLocation.latitude = view!!.currentLatitude
    geoLocation.longitude = view!!.currentLongitude

    if (response.main != null) {
      geoLocation.temperature = getFormattedCurrentTemp(response.main!!.temp!!)
      geoLocation.overAllTemperature = getFormattedOverallTemp(
          response.main!!.tempMin!!,
          response.main!!.tempMax!!
      )
    }
    if (response.weather != null && response.weather!!.isNotEmpty() && response.wind != null
    ) {
      geoLocation.description = getFormattedDescription(
          response.weather!![0].description,
          response.wind!!.speed!!
      )
    }
    insertLocationData(geoLocation)
  }

  private fun insertLocationData(geoLocation: GeoLocation) {
    view!!.locationDB?.geoLocationDao
        ?.insertLocation(geoLocation)
        ?.subscribeOn(Schedulers.io())
        ?.observeOn(AndroidSchedulers.mainThread())
        ?.subscribe(object : CompletableObserver {
          override fun onSubscribe(d: Disposable) {}
          override fun onComplete() {
            view!!.hideLoading()
            view!!.updateWeatherDetails(geoLocation)
          }

          override fun onError(e: Throwable) {
            view!!.hideLoading()
            view!!.updateWeatherDetails(geoLocation)
          }
        })
  }
}